ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -152,62 +161,72 @@
       );
       return forSource(
           builder,
+          topic.getKeyFormat(),
           StreamSourceBuilder.getKeySerde(builder, step),
           step,
-          keyField,
-          queryContext);
+          keyField);
     }
   }
 
-  protected SchemaKStream(
+  @VisibleForTesting
+  SchemaKStream(
       final KStream<K, GenericRow> kstream,
-      final LogicalSchema schema,
+      final ExecutionStep<KStream<K, GenericRow>> sourceStep,
+      final KeyFormat keyFormat,
       final KeySerde<K> keySerde,
       final KeyField keyField,
       final List<SchemaKStream> sourceSchemaKStreams,
       final Type type,
       final KsqlConfig ksqlConfig,
-      final FunctionRegistry functionRegistry,
-      final QueryContext queryContext
+      final FunctionRegistry functionRegistry
   ) {
     this(
-        kstream, schema,
+        kstream,
+        requireNonNull(sourceStep, "sourceStep"),
+        sourceStep.getProperties(),
+        keyFormat,
         keySerde,
         keyField,
         sourceSchemaKStreams,
         type,
         ksqlConfig,
         functionRegistry,
-        StreamsFactories.create(ksqlConfig),
-        queryContext);
+        StreamsFactories.create(ksqlConfig)
+    );
+    this.keyField.validateKeyExistsIn(getSchema());
   }
 
+  // CHECKSTYLE_RULES.OFF: ParameterNumber
   SchemaKStream(
       final KStream<K, GenericRow> kstream,
-      final LogicalSchema schema,
+      final ExecutionStep<KStream<K, GenericRow>> sourceStep,
+      final ExecutionStepProperties sourceProperties,
+      final KeyFormat keyFormat,
       final KeySerde<K> keySerde,
       final KeyField keyField,
       final List<SchemaKStream> sourceSchemaKStreams,
       final Type type,
       final KsqlConfig ksqlConfig,
       final FunctionRegistry functionRegistry,
-      final StreamsFactories streamsFactories,
-      final QueryContext queryContext
+      final StreamsFactories streamsFactories
   ) {
-    this.schema = requireNonNull(schema, "schema");
+    // CHECKSTYLE_RULES.ON: ParameterNumber
+    this.keyFormat = requireNonNull(keyFormat, "keyFormat");
     this.keySerde = requireNonNull(keySerde, "keySerde");
+    this.sourceStep = sourceStep;
+    this.sourceProperties = Objects.requireNonNull(sourceProperties, "sourceProperties");
     this.kstream = kstream;
-    this.keyField = requireNonNull(keyField, "keyField")
-        .validateKeyExistsIn(schema);
+    this.keyField = requireNonNull(keyField, "keyField");
```

Comment:
It's not actually dropped. This is another artifact of the misused inheritance here. This constructor gets used by SchemaKTable as well, which won't have set the source step yet, and so `getSchema()` would return null. So I moved the validation to the other constructor, which is used to create all the SchemaKStream instances, and the SchemaKTable constructor which is used to create all the SchemaKTable instances.

All that said, I think it would be more error-proof to just pass the schema in here.

Reply:
Actually never mind, we have the source properties here so we can just get the schema from there.